source/slang/slang-check-conversion.cpp

```diff
@@ -388,88 +429,246 @@ namespace Slang
+                    // 2. We have a `{}` (0 arguments), try to coerce
+                    // 3. We have a `{arg1, arg2...}`, try to coerce
 
-                    // No point in trying further if any argument fails
-                    if(!argResult)
-                        return false;
+                    auto toStructDecl = toStructDeclRef.getDecl();
+                    // 1. Ensure our `StructDecl` has declaratations created for all auto-generated constructors
+                    ensureDecl(toStructDecl, DeclCheckState::DefaultConstructorReadyForUse);
 
-                    if( coercedArg )
+                    // 2. We have a `{}` (0 arguments), try to coerce
+                    if (argCount == 0)
                     {
-                        coercedArgs.add(coercedArg);
+                        if (outToExpr)
+                        {
+                            *outToExpr = constructDefaultInitListFunc(this, toStructDecl, toType, ConstructDefaultInitListOptions::CheckToAvoidRecursion);
+                            (*outToExpr)->loc = fromInitializerListExpr->loc;
+                        }
+                        return true;
+                    }
+
+                    // 3. We have a `{arg1, arg2...}`, try to coerce
+                    //
+                    // For this situation we have more steps:
+                    // a. Find constructor candidate for our particular init-list
+                    // b. Collect and coerce init-list arguments into constructor parameters
+                    // c. Create InvokeExpr for our valid ConstructorDecl 
+                    List<ConstructorDecl*> ctorList = _getCtorList(this->getASTBuilder(), this, toStructDecl, nullptr);
+                    bool allowCStyleInitList = isCStyleStructDecl(this, toStructDecl, ctorList);
+
+                    List<Expr*> maybeArgList;
+                    maybeArgList.reserve(argCount);
+                    Index ioArgIndexMirror = ioArgIndex;
+                    UInt ioArgIndexCandidate = 0;
+
+                    for (auto& ctor : ctorList)
+                    {
+                        // a. Find constructor candidate for our particular init-list
+
+                        // Don't try to init default ctor with this logic
+                        auto ctorParamCount = ctor->getParameters().getCount();
+                        if (ctorParamCount == 0)
+                            continue;
+
+                        ioArgIndexCandidate = ioArgIndexMirror;
+                        ioArgIndex = ctorParamCount;
+
+                        // if allowCStyleInitList, process any ctor which comes next using the most members possible. ioArgIndex may not be 0.
+                        // if !allowCStyleInitList, process any ctor that exactly matched our argument count. ioArgIndex must start at 0.
+                        if (!allowCStyleInitList && ctorParamCount != Index(argCount))
```

Comment:
problem here is that in the case of a variadic ctor, the arg count will not match param count, because the param count is just 1 for the pack, that can match to many different args.